File: src/1146-Snapshot-Array/1146.py
class SnapshotArray:
    def __init__(self, length: int):
        self.dif = {}
        self.snaps = []

    def set(self, index: int, val: int) -> None:
        self.dif[index] = val

    def snap(self) -> int:
        self.snaps.append(self.dif.copy())
        return len(self.snaps) - 1

    def get(self, index: int, snap_id: int) -> int:
        return self.snaps[snap_id][index] if index in self.snaps[snap_id] else 0

File: src/1146-Snapshot-Array/1146.cpp
class SnapshotArray 
{
public:
    SnapshotArray(int length) {}
    
    void set(int index, int val) 
    {
        dif[index] = val;
    }
    
    int snap() 
    {
        snaps.push_back(dif);
        return snaps.size() - 1;
    }
    
    int get(int index, int snap_id) 
    {
        if (snaps[snap_id].count(index)) return snaps[snap_id][index];
        return 0;
    }
private:
    unordered_map<int, int> dif;
    vector<unordered_map<int, int>> snaps;
};